SimG4CMS/Calo/src/HFShowerLibrary.cc

```diff
@@ -19,58 +19,117 @@
 #include "CLHEP/Units/PhysicalConstants.h"
 
 //#define EDM_ML_DEBUG
+namespace {
+  HFShowerLibrary::Params paramsFrom(edm::ParameterSet const& hfShower,
+                                     edm::ParameterSet const& hfShowerLibrary,
+                                     double iDeltaPhi) {
+    HFShowerLibrary::Params params;
+
+    params.dphi_ = iDeltaPhi;
+
+    params.probMax_ = hfShower.getParameter<double>("ProbMax");
+    params.equalizeTimeShift_ = hfShower.getParameter<bool>("EqualizeTimeShift");
+
+    params.backProb_ = hfShowerLibrary.getParameter<double>("BackProbability");
+    params.verbose_ = hfShowerLibrary.getUntrackedParameter<bool>("Verbosity", false);
+    params.applyFidCut_ = hfShowerLibrary.getParameter<bool>("ApplyFiducialCut");
+
+    return params;
+  }
+
+  HFShowerLibrary::FileParams fileParamsFrom(edm::ParameterSet const& hfShowerLibrary) {
+    HFShowerLibrary::FileParams params;
+
+    edm::FileInPath fp = hfShowerLibrary.getParameter<edm::FileInPath>("FileName");
+    params.fileName_ = fp.fullPath();
+    if (params.fileName_.find('.') == 0)
+      params.fileName_.erase(0, 2);
+
+    std::string emName = hfShowerLibrary.getParameter<std::string>("TreeEMID");
+    std::string hadName = hfShowerLibrary.getParameter<std::string>("TreeHadID");
+    std::string branchEvInfo = hfShowerLibrary.getUntrackedParameter<std::string>(
+        "BranchEvt", "HFShowerLibraryEventInfos_hfshowerlib_HFShowerLibraryEventInfo");
+    std::string branchPre =
+        hfShowerLibrary.getUntrackedParameter<std::string>("BranchPre", "HFShowerPhotons_hfshowerlib_");
+    std::string branchPost = hfShowerLibrary.getUntrackedParameter<std::string>("BranchPost", "_R.obj");
+    params.fileVersion_ = hfShowerLibrary.getParameter<int>("FileVersion");
+
+    params.emBranchName_ = branchPre + emName + branchPost;
+    params.hadBranchName_ = branchPre + hadName + branchPost;
+
+    if (not branchEvInfo.empty()) {
+      params.branchEvInfo_ = branchEvInfo + branchPost;
+    }
+
+    params.cacheBranches_ = hfShowerLibrary.getUntrackedParameter<bool>("cacheBranches", false);
+    return params;
+  }
+}  // namespace
+
+HFShowerLibrary::HFShowerLibrary(const HcalDDDSimConstants* hcons,
+                                 const HcalSimulationParameters* hps,
+                                 edm::ParameterSet const& hfShower,
+                                 edm::ParameterSet const& hfShowerLibrary)
+    : HFShowerLibrary(paramsFrom(hfShower, hfShowerLibrary, hcons->getPhiTableHF().front()),
```

Comment:
Having a constructor that takes the two different ParameterSets allowed those values to be passed to two different functions without having to call `getParameterSet` several times.